### ISSUE TITLE
Ability for "Run on all agents" to run sequentially vs. in parallel

### ISSUE THREAD

**Issue by jefftrimm** [gocd/gocd#827]
Consider the use case of Go agents running on target application servers and executing deploy tasks.  
1. A resource on those agents running on the target application servers would be labelled with something like "[AppName]-Deploy".  
2. My deployment job then has "[AppName]-Deploy" defined as a required resource for the job.
3. I specify "Run on All Agents"
4. Go sends out the deployment tasks to all of the target deployment application servers in parallel.
5. All app servers will have their deployment jobs executed all at once.

However, if we are using this mechanism to orchestrate a production deployment, we don't want all nodes to be upgraded at once, creating an outage.  It would be very useful to have the execution for "Run on all agents" to have an option to run sequentially vs. in parallel.  Alternatively, it would be good to be able to specify "Run on all agents, x at one time".  Namely, run it across all matching agents, but run sequentially, with X agents running at the same time.  This would allow you to achieve faster throughput of deployment execution, while still managing how many target instances are taken out of service at once.

**Comment by hanej** [comment]
+1

This is preventing us from using this in production where we have several hundred applications that would all be whacked at the same time.

**Comment by jefftrimm** [comment]
There should also be an option to fail execution on all other agents if first execution fails or x% fails.  Consider deployment use-case using agents -- I want to deploy onto 20 servers in production.  If server 1 fails, I want to abort the entire deployment (not execute the other 19 agents).  Similarly, if the first deployment node succeeds, but then I see 15% deployment failure, I want to abort the rest of the deployments.  In the paranoid case, if I get a single failure, I may want to abort the rest of the deployments.

**Comment by JorritSalverda** [comment]
It would be also be useful to either use an integer for the number of agents to run at the same time, or a percentage to have the number automatically be calculated from the total number of suitable agents.

**Comment by jefftrimm** [comment]
To hopefully give this a nudge in terms of product prioritization -- this enhancement is HUGE in terms of making Go a viable framework to orchestrate deployments.  The advantage of being able to use Go end-to-end to have agents not only build/test code, but to have agents on the target hosts for deployment is the excellent end-to-end visibility you get in Go.  

For example, we also run smoke tests on the deployment hosts after a deploy is done.  Having those Xunit test results marked as test-resources and collected by Go is a huge benefit for troubleshooting.  IE. I deploy across 50 hosts, smoke tests fail on two of them.  Being to look at the Go console for the deploy pipeline and see the failed xunit test results for the failing nodes is a huge benefit.   It gives Go a nice advantage over other deployment frameworks.

But to take advantage of this, any organization that wants to be able to do a deploy without taking an outage needs the capability in this issue.  It is a must-have, not a nice-to-have, to use Go to orchestrate deployments.

**Comment by srinivasupadhya** [comment]
How about this. Divide the pipeline into stages (as many "serial" steps you want). So if you have 30 nodes & you want to deploy 10 at a time then make 3 stages all having the same job but each of them have one extra "resource" - `stage-1`, `stage-2` & `stage-3`. So you will have the corresponding 30 agents divided into 3 groups as well (using the above extra resource). This way when the first stage runs & goes green you can run smoke tests on them & then trigger the next stage that will deploy on next 10 which again will get tested separately & then the final 10.

**Comment by hanej** [comment]
While that may work, the issue is it's not a baked in solution.  We have to deploy to 200 servers for the same application and we can't be creating and maintains stages for that many servers.  Also anytime you scale the environment you have to go back and reorder your stages.  A parameter that defines how many servers in parallel it can deploy to would be a better solution.

**Comment by benzvan** [comment]
I made this suggestion quite a while ago. The response was that jobs are supposed to run in parallel. While I agree that was the philosophy within Go, it is a flawed philosophy for the reasons stated above. There should be a "run on all, one at a time" option.

**Comment by kmugrage** [comment]
This might be a hack...

If you had 5 agents tagged with the correct resources to do your production
deployment, there would be a max of 5 running at one time even if you tell
it to run on all agents.

In other words, you could tag the job with a resource like
"production_deploy" and then setup the number of agents you would want to
run at once with the same tag.

Ken

On Wed, Apr 15, 2015 at 10:22 AM, benzvan <notifications@github.com> wrote:

> I made this suggestion quite a while ago. The response was that jobs are
> supposed to run in parallel. While I agree that was the philosophy within
> Go, it is a flawed philosophy for the reasons stated above. There should be
> a "run on all, one at a time" option.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gocd/gocd/issues/827#issuecomment-93491903>.
>



-- 
Ken Mugrage - ThoughtWorks Studioskmugrage@thoughtworks.com+1(425) 835-3056
@kmugrage

**Comment by rahulghanate** [comment]
This is much-needed feature.
Can't use it for production services till we have sequential execution.
Currently creating separate environments for load balancing.
